### ISSUE TITLE
Can the prediction be made better?

### ISSUE THREAD

**Issue by punisherhaxball** [haxball/haxball-issues#1942]
Just curious about this.

https://www.youtube.com/watch?v=aJcRzZnQzKM

This is the first video ever taken about this thing and when I shared it with you, you said this has been in the game since 2010 and completely normal. I think it is not normal because there is prediction having an error and as the result we are seeing something not even happened. 

So the question is can't you make a better prediction where this fake kick not even happen single time? Because here in the video there is 0 extrapolation player (me) who seeing something never happened in reality, I think there must be a solution for that. If prediction having an error while predicting then why not improving it to avoid errors? Really wondering what you are thinking about this because thats making me curious for years.

**Comment by basro** [comment]
Even when you set extrapolation to 0 the game is still predicting some distance to compensate for the network latency (the ping). During prediction the game doesn't know what keys the players will be pressing or releasing in the future so it assumes the keys don't change. "Fake kicks" happen when a player would have kicked the ball in the next few frames had he kept his input the same, the player in fact released the kick button but the game doesn't know this during prediction (because this information has not yet reached your client).

I don't see any way to improve it.

**Comment by punisherhaxball** [comment]
I agree what you say but can't you not animate whatever not happened. I mean, if there is an animation happening based on something then force a rule on that animation to make sure if it's 100% happened kick.

**Comment by kingjasko** [comment]
> I agree what you say but can't you not animate whatever not happened. I mean, if there is an animation happening based on something then force a rule on that animation to make sure if it's 100% happened kick.

Wouldn't that just mean interpolation instead of extrapolation? Wouldn't it then cause input lag as well? Ideally we'd all be playing on a LAN and we wouldn't be facing such issues on 0 extrapolation, that being said, other games, even when not on LAN, don't really experience such drastic "prediction issues", so I wouldn't know. It's definitely a complex problem.

**Comment by punisherhaxball** [comment]
No what I am saying is he is drawing animations on screen based on some activity. So I am just saying do not draw animations if it's 100% confirmed action. Let people feel that input lag, atleast fix the solution.

**Comment by kingjasko** [comment]
If that's possible, i don't see why not, people could have it enabled or disabled perhaps, if they prefer how it is currently :D

**Comment by punisherhaxball** [comment]
Yeah exactly, I want to feel input lag as cost of removing warpyness and fake x. I can get used to it.
